src/kernel/lthread/lthread.c

```diff
@@ -39,37 +37,47 @@ static struct schedee *lthread_process(struct schedee *prev,
 	/* We have to restore ipl as soon as possible. */
 	ipl_enable();
 
-	lt->run_ret = next->run(next->run_arg);
+	lt->label_offset = lt->run(lt);
+
+	if (lt->joining && __lthread_is_disabled(lt))
+		sched_wakeup(lt->joining);
 
 	return NULL;
 }
 
-static void lthread_init(struct lthread *lt, void *(*run)(void *), void *arg) {
+void lthread_init(struct lthread *lt, int (*run)(struct lthread *)) {
 	assert(lt);
 
-	schedee_init(&lt->schedee, LTHREAD_PRIORITY_DEFAULT, lthread_process, run, arg);
+	schedee_init(&lt->schedee, SCHED_PRIORITY_HIGH, lthread_process);
 	sched_wait_info_init(&lt->info);
+
+	lt->run = run;
+	lt->joining = NULL;
+	lt->label_offset = 0;
 }
 
-struct lthread *lthread_create(void *(*run)(void *), void *arg) {
-	struct lthread *lt;
+void lthread_join(struct lthread *lt) {
+	struct schedee *self = schedee_get_current();
 
-	if (!run) {
-		return err_ptr(EINVAL);
-	}
+	assert(lt);
 
-	if (!(lt = (struct lthread *) pool_alloc(&lthread_pool))) {
-		return err_ptr(ENOMEM);
+	sched_wait_prepare();
```

Comment:
This macro uses the function from https://github.com/embox/embox/blob/master/src/kernel/thread/thread_sched_wait.c file that depends on some thread-specific headers and functions.